lib/trie/hash.go

```diff
@@ -17,31 +17,62 @@
 package trie
 
 import (
+	"bytes"
+	"context"
 	"hash"
+	"sync"
 
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/scale"
 	"golang.org/x/crypto/blake2b"
+	"golang.org/x/sync/errgroup"
 )
 
+type sliceBuffer []byte
+
+func (b *sliceBuffer) write(data []byte) {
+	*b = append(*b, data...)
+}
+
+func (b *sliceBuffer) reset() {
+	*b = (*b)[:0]
+}
+
 // Hasher is a wrapper around a hash function
 type Hasher struct {
-	hash hash.Hash
+	hash     hash.Hash
+	tmp      sliceBuffer
+	parallel bool // Whether to use parallel threads when hashing
+}
+
+// hasherPool creates a pool of Hasher.
+var hasherPool = sync.Pool{
+	New: func() interface{} {
+		h, _ := blake2b.New256(nil)
+
+		return &Hasher{
+			tmp:  make(sliceBuffer, 0, 520), // cap is as large as a full branch node.
+			hash: h,
+		}
+	},
 }
 
 // NewHasher create new Hasher instance
-func NewHasher() (*Hasher, error) {
-	h, err := blake2b.New256(nil)
-	if err != nil {
-		return nil, err
-	}
+func NewHasher(parallel bool) *Hasher {
+	h := hasherPool.Get().(*Hasher)
+	h.parallel = parallel
+	return h
+}
 
-	return &Hasher{
-		hash: h,
-	}, nil
+func returnHasherToPool(h *Hasher) {
```

Comment:
make this a receiver function on `Hasher`? It would be nice to just be able to `defer h.Return()`.

Reply:
Done